### ISSUE TITLE
Cup-kampen mot MIF tolkes som seriekamp

### ISSUE THREAD

**Issue by havremunken** [RosenborgSupporterSoftware/staut#13]
Jeg har ting gående på kottserveren min, og har fått filer for begge kampene mot bruntrøyene.

2015_LEAGUE_11_MjÃ¸ndalen_438529
2015_LEAGUE_13_MjÃ¸ndalen_468249

Jeg ANTAR at dette skyldes måten Billettservice navngir eventen på, men så er spørsmålet - klarer vi å identifisere en cupkamp automatisk? Eller må vi inn og gjøre manuelle tiltak når sånt skjer?

**Comment by vemundo** [comment]
Hm, jeg må se nærmere på hva som er merket i eventdetails for denne kampen,
jeg trodde jeg hadde laget en løsning som ville virke for NM også.

2015-07-27 15:28 GMT+02:00 Rune Jacobsen notifications@github.com:

**Comment by havremunken** [comment]
Et par andre ting jeg ser nå som jeg er innom en tur;
- Den har tilsynelatende ikke plukket opp kampen mot Debrechen her hos meg.
- vif og den ene MIF-kampen (438529) blir stadig oppdatert. Den andre MIF-kampen (468249) er ikke oppdatert siden 28/7 kl. 00:42. Da kom det en fil på 4k.

Vet ikke om noen av disse er indikasjon på noen krise, men.. :)

**Comment by vemundo** [comment]
Tilbake i vanlig hverdag igjen nå. Ser på problemet, men usikker på hva det er som skjer. I praksis får jeg ikke lest fra html-siden for cupkampen mot Mjøndalen fra java koden. Dvs den leser tilsynelatende fra URLen og det er riktig URL men den har 0 linjer.

Må grave i dette.

Når jeg finner ut av det skal jeg sjekke hvorfor cupkampen ikke ble klassifisert riktig når nedlastingen virket.

**Comment by havremunken** [comment]
Vel overstått! :)

Satt og så litt på det nå jeg også etter aktivitet i tråden på RBKweb. Dette er sikkert ikke noe du ikke allerede vet, men bare i tilfelle det har noen verdi:

Cupkampen sin availabilityUrl er https://www.billettservice.no/seatmap/proxy/availability/NO/TLD1308N-2015,NO-468249.xml. Fant denne på "gamlemåten" med view source og søke på availabilityUrl. Men collectoren ser ut til å være misfornøyd:

```
PS E:\staut\java> & 'C:\Program Files\Java\jre1.8.0_45\bin\java.exe' -jar .\dist\lib\STAut.jar --collector
INFO: Loading configuration from E:\staut\java\dist\lib\configuration.properties
INFO: Updating event collections.
INFO: Failed to find availabilityURL for event eventId: 468249
INFO: Most likely internet sale has not yet started.
```

Det vet vi jo er feil, spørsmålet er jo bare hvorfor vi ikke ser den når vi ser det på de andre. 

Her hos meg klarte den å laste ned data frem til 3/8 kl. 23:42 - ingenting for denne kampen etter det. Har ikke peiling på hva som stoppet opp, dog.

Vet ikke hvordan vegardj gjør nedlastning for sin graf, men den er visst heller ikke helt up to date, så kanskje noe trøbbel der også.

**Comment by vemundo** [comment]
Har funnet ut av problemet. 

De har på ett eller annet tidspunkt gått over til https (sannsynligvis den natta det begynte å feile) istedet for http for salgssidene sine. Men bare for cup-kamper virker det som. Dersom du brukere HTTP url istedet for HTTPS så vil en browser gjøre automatisk redirect for deg, men min java kode var ikke så smart.

Jeg har lagt til mer feilsjekking nå og leser også ut redirect informasjon fra response headeren på HTTP requesten, så jeg kan automatisk gjøre et nytt oppslag på den siden det redirigeres til. Dette ser ut til å virke. Var bare et problem for salgssiden, men har oppdatert koden for oppkobling mot billettservice sin søkeside og selve availabilityURL osv. til å gjøre det samme så vi ikke plutselig blir overrasket av tilsvarende endringer. 

Lagrer nå også den endelig URLen som vi ender opp med å bruke i eventinfo fila så vi vet det for fremtiden. Skriver en del mer ut når du starter opp collector modusen, så litt mer støy men tror det kan hjelpe om vi treffer andre problemer rundt oppkobling og nedlasting.

Har pushet endringene til git.

**Comment by havremunken** [comment]
Veldig bra, ser ut til å funke!

Spesielt og sneaky av BS å gjøre dette kun med NM-kampene. Kanskje de tester oss. ;)

Nå ser det forøvrig også ut som om cup-kampen blir oppfattet som NM. Helt glimrende!

**Comment by vemundo** [comment]
Ja, jeg tror den fiksen med NM var noe jeg hadde fikset og pushet før ferien men ikke varslet deg om.

Det eneste som ikke er helt topp der er at de ikke inkluderer hvilken "runde" det er i cupen som en del av koden sin, istedet har de bare bakt in datoen (1308). Så, får ikke hentet ut at det er "kvartfinale" eller runde 5(?) på noen trygg måte. Men det er bakt inn i navnet på eventen da, som kan evt brukes som en header for grafen f.eks.

**Comment by havremunken** [comment]
Å klare å finne frem til "runde" i NM bør jo være mulig basert på dato-range, sånn ca., bortsett fra at det vel også forekommer utsatte kamper der. Dette finner vi ut av i takt med erfaring med flere kamper. Jeg håper ihvertfall det blir flere kamper i årets NM... ;) Then again, sjansen er jo god for at semifinalen ikke spilles på Lerkendal uansett.

**Comment by vemundo** [comment]
I verste fall kan vi evt hente data om "runde" fra en annen webside med mer detaljerte data om kampene og matche mot informasjonen vi har om kampens dato og klokkeslett fra BS.

**Comment by havremunken** [comment]
Det bør jo gå an ja!

Collectoren fant tilsynelatende ikke automatisk kampen mot Steaua i går når den ble lagt ut for salg. Jeg hadde den gående fra Debrechen-kampen ble lagt ut og til godt etter at den var spilt, og så aldri noen filer for den. Når jeg restartet i går, fant den Steaua og begynte å laste ned.

Er det noe som kan være verdt å ta en titt på?

**Comment by vemundo** [comment]
Hm ja, det er jo ikke sånn den skal fungere. Det er samme koden som kjøres når den startes opp som skal kjøres by nytt 2 ganger i døgnet eller noe sånt for å se etter nye kamper. Må prøve å finne en elegant måte jeg kan simulere den situasjonen på.

**Comment by havremunken** [comment]
Bare slenger inn en kjapp kommentar her - i dag telte den vif siste gang kl. 17:33. Det er forsåvidt forståelig nok, 18:03 var jo kampen i gang... Men jeg skulle gjerne ha mulighet til å telle helt til 2. omgang starter, fordi erfaringsmessig er det først da (eller ihvertfall et stykke inn i første) at tallene våre stemmer med det RBK offentliggjør.

Deilig seier i dag! :grin:

**Comment by vemundo** [comment]
Logikken er slik at hver X timer (12?) sjekker collectoren hvilke kamper som dukker opp fra søkesiden til Billettservice. Dersom det er noen nye kamper starter den en tråd som står og samler availability.xml for denne kampen. Dersom det er en aktiv kamp som ikke lenger synes på oversikten så stopper collectoren den tråden som sto og samlet data for denne kampen.

Hm, kan tyde på litt uheldig timing her, at 12-timers sjekken slo inn rett før/etter kampstart og at kampen da ikke lenger var oppe på BS sine sider. Vet ikke om vi skal endre på logikken her. Vi kan jo også endre på hvor ofte vi sjekker etter nye kamper og avsluttede kamper. evt. bruke faste klokkeslett for det. Vet ikke.

**Comment by vemundo** [comment]
Klarer ikke reprodusere dette problemet. Er det mulig at før collectoren ble restarted så kjørte den fremdeles med den versjonen som ikke håndterte http redirects, f.eks. til/fra http/https? 

Hver gang collectoren (hver 12. time) sjekker etter nye kamper og om eksisterende salg er over, skriver den ut:

"Updating event collections."

Husker du om dette ble skrevet ut hver 12. time? Hvis ikke kan det hende den tråden som kjører oppdateringene hadde trynet pga en bug av noe slag.

Kjørte en runde med sjekkekoden hvert minutt nå i noen timer uten å treffe noen problemer.

Forresten, om du vil sjekke oftere eller sjeldnere enn hver 12. time kan du endre denne i configuration.properties:

check.events.period=PT12H

Etterpå må du bygge med ant for at endringen skal ha effekt neste gang du starter collectoren.

PS. Hvis du bygger mens collectoren kjører vil den etterhvert crashe, det vil skrive over jar-fila under beina på den.

**Comment by havremunken** [comment]
Siden jeg er så lite kjent med java-byggeprosessen stoppet jeg alt, pullet fra git, kjøre clean og build, og kjørte på nytt. Men skal åpne for at jeg kan ha gjort en verden av feil her. :)

Skal se på dette når jeg kommer hjem og har litt pusterom. Bør ha logg i et konsollvindu, så kan vi se om vi skjønner hva som skjer.

**Comment by havremunken** [comment]
Pga strømbrudd var det lite logg å få til forrige uke dessverre.

Bare for å prøve litt, satt jeg tiden for collecting til 10 minutter nå (ned fra 30). Likevel sa loggen:

INFO: Starting collection of URL: http://www.billettservice.no/seatmap/proxy/ava
ilability/NO/TLD1215-2015,NO-438531.xml every 30 minutes.

Har sittet og ventet i 10m nå, og det skjer ikke noe.

Sånn ser config-fila ut:

```
archive.directory=./archive/
check.events.period=PT6H
collect.availability.period=PT10M
download.retry.count=3
download.retry.wait=PT1M
```

Må jeg faktisk recompile osv. for at config skal tas hensyn til, eller leser den fila til runtime?

**Comment by vemundo** [comment]
Ja, du må dessverre kjøre build-stegene, da kopieres configuration.properties inn under dist/lib/ sammen med STAut.jar. Programmet forventer å finne konfig-fila i samme katalog som jar-fila. Vi kan jo endre på dette. Evt. kan vi plassere STAut.jar i samme katalog som du står i når du bygger så ligger configuration.properties der allerede.

Hvis du kjører 

java -jar dist/lib/STAut.jar --config

så printer den ut den konfigurasjonen den bruker.
